README.md
[![Rust](https://github.com/rodrimati1992/arrcat/workflows/Rust/badge.svg)](https://github.com/rodrimati1992/arrcat/actions)
[![crates-io](https://img.shields.io/crates/v/arrcat.svg)](https://crates.io/crates/arrcat)
[![api-docs](https://docs.rs/arrcat/badge.svg)](https://docs.rs/arrcat/*)


Array concatenation

This crate allows concatenating multiple arrays of varying lengths into one array.

# Example

### `concat_arrays`

For more examples of using [`concat_arrays`],
you can [look here][concat_arrays_examples].

```rust
use arrcat::concat_arrays;

{
    const PRIMES: [u16; 4] = [7, 11, 13, 17];
    assert_eq!(
        concat_arrays!([3, 4, 4u16.pow(3)], PRIMES),
        [3, 4, 64, 7, 11, 13, 17],
    );
}

{

    let increasing = [8, 9, 10];

    let concated = concat_arrays!(
        // the macro can't infer the length of runtime array non-literals.
        increasing: [_; 3],
        // most non-literal arguments need to be wrapped in `()` or `{}`.
        ([2u16, 3, 4].map(|x| x * 9)): [_; 3],
    );

    assert_eq!(concated, [8, 9, 10, 18, 27, 36]);
}

```

# No-std support

`arrcat` is `#![no_std]`, it can be used anywhere Rust can be used.

# Minimum Supported Rust Version

`arrcat` requires Rust 1.57.0, requiring crate features to use newer language features.


[`concat_arrays`]: https://docs.rs/arrcat/*/arrcat/macro.concat_arrays.html
[concat_arrays_examples]: https://docs.rs/arrcat/*/arrcat/macro.concat_arrays.html#examples


Changelog.md
This is the changelog, summarising changes in each version(some minor changes may be ommited).

# 0.1

### 0.1.0

Defined the `concat_arrays` macro.

src/tests.rs
use crate::concat_arrays;

use core::{cmp::PartialEq, fmt::Debug};

macro_rules! generic_test {
    ($conv:ident => $($code:tt)*) => {
        fn inside<T>()
        where
            T: Copy + From<u16> + PartialEq + Debug
        {
            let $conv = <T as From<u16>>::from;
            $($code)*
        }

        inside::<u16>();
    };
}

#[test]
fn test_empty_array() {
    generic_test! {_f =>
        {
            asserteq(concat_arrays!(), [0u8; 0]);
        }
        {
            asserteq(concat_arrays!([]), [0u8; 0]);
        }
        {
            let arr = concat_arrays!([]: [T; 0]);
            asserteq(arr, []);
        }
        {
            let arr = concat_arrays!([]: [T; _]);
            asserteq(arr, []);
        }
    }
}

#[test]
fn test_array_arg() {
    generic_test! {f=>
        // elems
        {
            let arr = concat_arrays!([f(3), f(4)]);
            asserteq(arr, [f(3), f(4)]);
        }
        // repeated, transparent type ascription
        {
            let arr = concat_arrays!([f(3); 2]: [T; 2]);
            asserteq(arr, [f(3), f(3)]);
        }
        // repeated, transparent type ascription, inferred length
        {
            let arr = concat_arrays!([f(3); 2]: [T; _]);
            asserteq(arr, [f(3), f(3)]);
        }
        // repeated, opaque type ascription
        {
            let arr = concat_arrays!([f(3); 2]: ([T; 2]));
            asserteq(arr, [f(3), f(3)]);
        }
    }
}

#[test]
fn test_const_arg() {
    // constant
    {
        const B: D = D(10);
        const C: [D; 2] = [B; 2];
        let arr = concat_arrays!(C);
        asserteq(arr, [D(10), D(10)]);
    }
    // constant, transparent type ascription
    asserteq(concat_arrays!(FooConst::C: [u8; 4]), [3, 3, 3, 3]);
    asserteq(concat_arrays!((FooConst::C): [u8; 4]), [3, 3, 3, 3]);

    // constant, transparent type ascription, inferred length
    asserteq(concat_arrays!(FixedLengthArray::C: [u32; _]), [5, 5, 5]);
    asserteq(concat_arrays!((FixedLengthArray::C): [u32; _]), [5, 5, 5]);

    // constant, opaque type ascription
    {
        type T = [u16; 2];
        let arr = concat_arrays!((FooConst::C): T);
        asserteq(arr, [4, 4]);
    }
}

#[test]
fn test_runtime_variable_arg() {
    // variable, transparent type ascription
    {
        let var = FooConst::C;
        let arr = concat_arrays!(var: [u8; 4]);
        asserteq(arr, [3, 3, 3, 3]);
    }
    // variable, opaque type ascription
    {
        type T = [u16; 2];
        let var = FooConst::C;
        let arr = concat_arrays!(var: T);
        asserteq(arr, [4, 4]);
    }

    // variable in parens, transparent type ascription
    {
        let var = FooConst::C;
        let arr = concat_arrays!((var): [u8; 4]);
        asserteq(arr, [3, 3, 3, 3]);
    }
    // variable in parens, opaque type ascription
    {
        type T = [u16; 2];
        let var = FooConst::C;
        let arr = concat_arrays!((var): T);
        asserteq(arr, [4, 4]);
    }
}

#[test]
fn test_runtime_expression_arg() {
    fn do_nothing() {}
    fn rt<T>(x: T) -> T {
        x
    }

    // runtime expression, transparent type ascription, expr in braces
    {
        let arr = concat_arrays!(
            {
                do_nothing();
                rt(FooConst::C)
            }: [u8; 4]
        );
        asserteq(arr, [3, 3, 3, 3]);
    }
    // runtime expression, transparent type ascription, expr in parentheses braces
    {
        let arr = concat_arrays!((rt(FooConst::C)): [u8; 4]);
        asserteq(arr, [3, 3, 3, 3]);
    }
    // runtime expression, opaque type ascription
    {
        type T = [u16; 2];
        let arr = concat_arrays!((rt(FooConst::C)): T);
        asserteq(arr, [4, 4]);
    }
}

#[test]
fn test_path_args() {
    {
        let arr = concat_arrays!(::core_extensions::ConstDefault::DEFAULT: [u8; 3]);
        asserteq(arr, [0, 0, 0]);
    }
    {
        let arr = concat_arrays!(core_extensions::ConstDefault::DEFAULT: [Option<Str>; 3]);
        asserteq(arr, [None, None, None]);
    }
}

mod arrays {
    pub const ARR0: [u16; 3] = [16, 32, 64];
}

#[test]
fn test_many_args() {
    {
        const C: [u16; 3] = [21, 34, 55];

        let var = [89, 144, 233];

        let arr = concat_arrays!([3, 5, 8], [13; 2], self::arrays::ARR0, (C), (var): [_; 3],);
        asserteq(arr, [3, 5, 8, 13, 13, 16, 32, 64, 21, 34, 55, 89, 144, 233]);
    }
}

/// Tests all kinds of arguments with:
/// - interspersed commas
/// - trailing commas
#[test]
fn test_comma_sep() {
    fn rt<T>(x: T) -> T {
        x
    }

    const RHS: [u16; 5] = [3, 5, 8, 13, 13];
    type Arr2 = [u16; 2];
    type Arr3 = [u16; 3];

    // arrays, no type ascription
    asserteq(concat_arrays!([3, 5, 8], [13, 13]), RHS);
    asserteq(concat_arrays!([3, 5, 8], [13; 2],), RHS);
    asserteq(concat_arrays!([3, 5, 8], [13; 2], []), RHS);

    // arrays, with transparent type ascription
    asserteq(concat_arrays!([3, 5, 8]: [_; 3], [13; 2]: [_; 2]), RHS);
    asserteq(concat_arrays!([3, 5, 8]: [_; 3], [13, 13]: [_; 2],), RHS);
    asserteq(concat_arrays!([3, 5, 8]: [_; 3], [13, 13]: [_; 2], []), RHS);

    // arrays, with opaque type ascription
    asserteq(concat_arrays!([3, 5, 8]: Arr3, [13, 13]: Arr2), RHS);
    asserteq(concat_arrays!([3, 5, 8]: Arr3, [13; 2]: Arr2,), RHS);
    asserteq(concat_arrays!([3, 5, 8]: Arr3, [13; 2]: Arr2, []), RHS);

    const B: [u16; 2] = [13, 13];
    // expression, no type ascription
    asserteq(concat_arrays!([3, 5, 8], (B)), RHS);
    asserteq(concat_arrays!([3, 5, 8], (B),), RHS);
    asserteq(concat_arrays!([3, 5, 8], B), RHS);
    asserteq(concat_arrays!([3, 5, 8], B,), RHS);
    asserteq(concat_arrays!([3, 5, 8], B, []), RHS);

    // expression, with transparent type ascription
    asserteq(concat_arrays!([3, 5, 8], (rt(B)): [u16; 2]), RHS);
    asserteq(concat_arrays!([3, 5, 8], (rt(B)): [u16; 2],), RHS);
    asserteq(concat_arrays!([3, 5, 8], B: [u16; 2]), RHS);
    asserteq(concat_arrays!([3, 5, 8], B: [u16; 2],), RHS);
    asserteq(concat_arrays!([3, 5, 8], B: [u16; 2], []), RHS);

    // expression, with opaque type ascription
    asserteq(concat_arrays!([3, 5, 8], (rt(B)): Arr2), RHS);
    asserteq(concat_arrays!([3, 5, 8], (rt(B)): Arr2,), RHS);
    asserteq(concat_arrays!([3, 5, 8], B: Arr2), RHS);
    asserteq(concat_arrays!([3, 5, 8], B: Arr2,), RHS);
    asserteq(concat_arrays!([3, 5, 8], B: Arr2, []), RHS);
}

#[test]
fn test_macro_called_by_macro() {
    {
        macro_rules! prepended_lit {
            ($prefix:tt) => {
                concat_arrays!($prefix, ["hello", "world"])
            };
        }

        // parsed as array by macro
        asserteq(
            prepended_lit!(["foo", "bar"]),
            ["foo", "bar", "hello", "world"],
        );
        asserteq(prepended_lit!(["foo"; 2]), ["foo", "foo", "hello", "world"]);

        // constant, not parsed as array by macro
        asserteq(
            prepended_lit!((["foo", "bar"])),
            ["foo", "bar", "hello", "world"],
        );
        asserteq(
            prepended_lit!((["foo"; 2])),
            ["foo", "foo", "hello", "world"],
        );
    }

    {
        let array = [3, 5, 8];

        macro_rules! prepended {
            ($prefix:tt : $ty:tt) => {
                concat_arrays!($prefix: $ty, [80, 81])
            };
        }

        asserteq(prepended!([3, 5, 6]: _), [3, 5, 6, 80, 81]);
        asserteq(prepended!([3, 5, 6]: [_; _]), [3, 5, 6, 80, 81]);
        asserteq(prepended!(array: [_; 3]), [3, 5, 8, 80, 81]);
    }
}

#[test]
fn length_type_arg() {
    {
        enum L {}
        let _: &[u8] = &concat_arrays!(length_type = L;);
        assert_eq!(L::LEN, 0);
    }
    {
        enum L {}
        let _: &[u8] = &concat_arrays!(length_type = L; [1]);
        assert_eq!(L::LEN, 1);
    }
    {
        const C: [u8; 5] = [5, 8, 13, 21, 34];
        enum L {}
        const A: [u8; L::LEN] = concat_arrays!(length_type = L; [1], [2; 3], C);
        asserteq(A, [1, 2, 2, 2, 5, 8, 13, 21, 34]);
        assert_eq!(L::LEN, 9);
    }
}

#[derive(Debug, PartialEq)]
struct Str(&'static str);

impl ::core_extensions::ConstDefault for Str {
    const DEFAULT: Self = Self("");
}

impl Drop for Str {
    fn drop(&mut self) {}
}

trait ConstVal<T> {
    const C: T;
}

struct FooConst;

impl ConstVal<u8> for FooConst {
    const C: u8 = 3;
}

impl ConstVal<u16> for FooConst {
    const C: u16 = 4;
}

impl ConstVal<u32> for FooConst {
    const C: u32 = 5;
}

impl<T, const L: usize> ConstVal<[T; L]> for FooConst
where
    Self: ConstVal<T>,
{
    const C: [T; L] = [<FooConst as ConstVal<T>>::C; L];
}

struct FixedLengthArray;

impl<T> ConstVal<[T; 3]> for FixedLengthArray
where
    FooConst: ConstVal<T>,
{
    const C: [T; 3] = [<FooConst as ConstVal<T>>::C; 3];
}

#[derive(Debug, PartialEq)]
struct D(u32);

impl Drop for D {
    fn drop(&mut self) {}
}

#[track_caller]
fn asserteq<T, const L: usize, const R: usize>(found: [T; L], expected: [T; R])
where
    T: PartialEq + Debug,
{
    assert_eq!(&found[..], &expected[..]);
}


src/internals.rs
use core::{marker::PhantomData, mem::ManuallyDrop};

#[doc(hidden)]
#[derive(Copy, Clone)]
pub struct Usize<const N: usize>;

impl<const N: usize> Usize<N> {
    #[inline(always)]
    pub const fn infer_mda<T>(self, _: &ManuallyDrop<[T; N]>) {}

    #[inline(always)]
    pub const fn get(self) -> usize {
        N
    }
}

///
/// # Safety
///
/// Implementors must have exactly one type parameter,
/// and the `T` associated type must be the value of that type parameter.
#[doc(hidden)]
pub unsafe trait GetTypeParam: Sized {
    type T;

    const PROOF: TypeParam<Self, Self::T> = TypeParam { types: PhantomData };
}

pub struct TypeParam<S, T> {
    types: PhantomData<fn() -> (PhantomData<S>, PhantomData<T>)>,
}

impl<S, T> Copy for TypeParam<S, T> {}

impl<S, T> Clone for TypeParam<S, T> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<S, T> TypeParam<S, T> {
    #[inline(always)]
    pub const fn assert_type_param(self, _: PhantomData<T>) {}

    pub const unsafe fn new_unchecked() -> Self {
        Self::__NEW
    }

    const __NEW: Self = Self { types: PhantomData };
}

pub unsafe trait ArrayLength {
    const LENGTH: usize;
}

unsafe impl<T, const L: usize> GetTypeParam for [T; L] {
    type T = T;
}

unsafe impl<T, const L: usize> ArrayLength for [T; L] {
    const LENGTH: usize = L;
}

#[repr(transparent)]
pub struct Identity<T> {
    pub inner: T,
}

#[repr(transparent)]
pub struct ArrayAndGhost<T, const LEN: usize> {
    pub inner: [T; LEN],
    pub elem_ty: PhantomData<T>,
}

#[doc(hidden)]
pub const unsafe fn concat_arrays<From_, T, const CONCAT_LEN: usize>(
    this: From_,
    _param: TypeParam<From_, T>,
) -> [T; CONCAT_LEN] {
    use core::mem::size_of;

    assert!(size_of::<From_>() == size_of::<[T; CONCAT_LEN]>());

    const_transmute!(
        ArrayHList<Rem, T, LEN>,
        [T; CONCAT_LEN],
        this
    )
}

/// Helper type for transmuting non-Copy types without adding any overhead in debug builds.
///
#[doc(hidden)]
#[repr(C)]
pub union TransmuterMD<T, U> {
    pub from: ManuallyDrop<T>,
    pub to: ManuallyDrop<U>,
}

macro_rules! const_transmute {
    ($from:ty, $to:ty, $val:expr) => {
        $crate::__::ManuallyDrop::into_inner(
            $crate::__::TransmuterMD {
                from: $crate::__::ManuallyDrop::new($val),
            }
            .to,
        )
    };
}
use const_transmute;


src/macros/concat_array_macro.rs
/// Concatenates multiple arrays into one.
///
/// This macro is callable in const contexts.
///
/// [**examples below**](#examples)
///
/// # Syntax
///
/// The syntax of this macro, using `macro_rules!`-like input syntax
///
/// ```text
/// concat_arrays!{
///     $( length_type = $length_type:ty ;)?
///     
///     $( $array_arg:array_expr $(: $argument_type:ty )? ),*
///     $(,)?
/// }
/// ```
///
/// Where `$length_type` is a concrete type.
/// When this argument is passed,
/// a `$length_type::LEN` associated constant
/// is defined with the length of the returned array.
/// [example below](#length-inference-example)
///
/// Where `$array_arg` can be any of:
///
/// - `[ $($array_contents:tt)* ]`: an array literal.
///
/// - A `()`/`{}`-delimited expression of array type. Eg: `(foo.bar())`, `{foo(); bar()}`.
///
/// - `$path:path` expression of array type. Eg: `foo`, `::foo::bar`, `Foo::<T>::BAR`.
///
/// Where `$argument_type` is the type of that argument (always an array).
///
/// ### Special syntax
///
/// This macro allows inferring the length of arrays in type annotations, with a `_` length:
/// ```rust
/// # use arrcat::concat_arrays;
/// # assert_eq!(
/// // The `_` tells this macro to infer the length.
/// concat_arrays!([3, 5, 8]: [u8; _])
/// #   [..],
/// #   [3, 5, 8]
/// # );
/// ```
///
/// # Limitations
///
/// This macro cannot concatenate arrays whose length depends on a
/// surrounding generic parameter.
///
/// These are the only kinds of arguments that don't require a type annotation:
///
/// - Constants of fully inferred type.
///
/// - Array literals (they can contain runtime values).
///
/// ### Argument from other macros
///
/// Note that due to how `:expr` macro parameters work,
/// arguments of that type from other macros aren't parsed as an array literal,
/// so they may require a type annotation.
///
/// These are two ways that the caller macro can parse that parameter to
/// pass it as an array literal:
/// - `$array_arg:tt`, passed as `$array_arg`
/// - `[$($array_arg:tt)*]`, passed as `[$($array_arg)*]`
///
/// The same limitation applies to parsing type annotations.<br>
/// If the caller macro passes a `$type:ty` as the type of an argument,
/// it'll require the length of the array to be specified.
///
/// # Examples
///
/// ### Basic
///
/// ```rust
/// use arrcat::concat_arrays;
///
/// assert_eq!(
///     concat_arrays!(["foo", "bar"], ["baz"; 3]),
///     ["foo", "bar", "baz", "baz", "baz"],
/// );
/// ```
///
/// ### Constant arguments
///
/// ```rust
/// use arrcat::concat_arrays;
///
/// assert_eq!(consts(), [Some(3), Some(5), None, Some(8), None, None, None]);
///
/// fn consts() -> [Option<u16>; 7] {
///     concat_arrays!(
///         [Some(3), Some(5)],
///         // `const`ants can passed unparenthesized.
///         FOO,
///         // Constant array args must have a fully inferred type.
///         // Since `Bar::ARR` isn't fully inferred, it requires a type annotation.
///         Bar::ARR: [Option<u16>; _],
///     )
/// }
///
/// const FOO: [Option<u16>; 2] = [None, Some(8)];
///
/// struct Bar<T>(T);
///
/// impl<T> Bar<T> {
///     pub const ARR: [Option<T>; 3] = [None, None, None];
/// }
///
/// ```
///
/// ### Non-constant arguments
///
/// ```rust
/// use arrcat::concat_arrays;
///
/// assert_eq!(VARS, [12, 13, 14, 24, 28, 32, 243]);
///
///
/// const VARS: [u16; 7] = {
///     let twelve = 12;
///     let variable = [24, 28];
///
///     concat_arrays!(
///         [twelve, twelve + 1, twelve + 2],
///         // Array variables can passed unparenthesized, and require type annotations.
///         // Non-constant arguments require type annotation
///         // for this macro to know their array length.
///         variable: [_; 2],
///         // Other expressions need to be parenthesized.
///         (make_array(5)): [_; 2],
///     )
/// };
///
/// const fn make_array(x: u32) -> [u16; 2] {
///     [2u16.pow(x), 3u16.pow(x)]
/// }
/// ```
///
/// <span id = "length-inference-example"></span>
/// ### Length constant
///
/// This macro allows getting the length of the returned array,
/// by declaring an associated constant on a passed-in type.
///
/// The length can then be used anywhere that can access the passed-in type.
///
/// ```rust
/// use arrcat::concat_arrays;
///
/// assert_eq!(foo(0xF00), [0xF00, 3, 5, 8, 13, 21, 34, 55]);
///
/// enum FooLen{}
/// const fn foo(x: u64) -> [u64; FooLen::LEN] {
///     let foo = [x, 3, 5, 8];
///
///     concat_arrays!{
///         // makes the macro define the `FooLen::LEN` associated constant with
///         // the length of the returned array
///         length_type = FooLen;
///         
///         foo: [_; 4],
///         [13, 21, 34, 55],
///     }
/// }
///
/// ```
#[macro_export]
macro_rules! concat_arrays {
    () => ([]);
    ($(length_type = $length_type:ty)?; $($args:tt)* ) => (
        $crate::__concat_arrays_preprocess_inner!{
            (config(length_type($($length_type)?)))
            ($($args)*)
        }
    );
    ( $($args:tt)* ) => (
        $crate::__concat_arrays_inner!{(config(length_type())) ($($args)*)}
    );
}

#[doc(hidden)]
#[macro_export]
macro_rules! __concat_arrays_preprocess_inner {
    ( $config:tt () ) => {
        $crate::__concat_arrays_inner! {$config ([])}
    };
    ( $config:tt $args:tt ) => {
        $crate::__concat_arrays_inner! {$config $args}
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! __concat_arrays_inner {
    (
        (
            config(length_type $length_type:tt)

            $(
                (
                    $expr:expr,
                    ($($elem:ty)?),
                    ($len:expr),
                    ($($type:tt)*),
                )
            )*
        )
        ($(,)?)
    ) => {
        unsafe{
            #[repr(C, packed)]
            struct __Concater<__PrivT>(
                $([__PrivT; $len],)*
            );

            impl<__PrivT> __Concater<__PrivT> {
                const PROOF: $crate::__::TypeParam<Self, __PrivT> = unsafe{
                    $crate::__::TypeParam::new_unchecked()
                };
            }

            $crate::__::concat_arrays::<
                _,
                _,
                {$crate::__declare_length_type_and_pass!(
                    $length_type,
                    {
                        let mut len = 0;
                        $( len += $len; )*
                        len
                    }
                )}
            >(
                __Concater(
                    $($crate::__type_ascription!(($expr) ($($type)*)),)*
                ),
                __Concater::PROOF,
            )
        }
    };

    (
        ($($prev:tt)*)
        ( [$($array:tt)*] $(: [$elem_ty:ty; $($len:tt)*])?  $(, $($rem:tt)*)? )
    ) => {
        $crate::__concat_arrays_inner!{
            (
                $($prev)*
                (
                    [$($array)*],
                    ($($elem_ty)?),
                    ($crate::__get_array_length!($($array)*)),
                    ($([$elem_ty; $($len)*])?),
                )
            )
            ($($($rem)*)?)
        }
    };

    (
        ($($prev:tt)*)
        ( [$($array:tt)*] $(: $type:ty)?  $(, $($rem:tt)*)? )
    ) => {
        $crate::__concat_arrays_inner!{
            (
                $($prev)*
                (
                    [$($array)*],
                    ($(<$type as $crate::__::GetTypeParam>::T)?),
                    ($crate::__get_array_length!($($array)*)),
                    ($($type)?),
                )
            )
            ($($($rem)*)?)
        }
    };


    (
        ($($prev:tt)*)
        ( $expr:tt $(: [$elem_ty:ty; $($len:tt)*])?  $(, $($rem:tt)*)? )
    ) => {
        $crate::__concat_arrays_inner!{
            (
                $($prev)*
                (
                    $expr,
                    ($($elem_ty)?),
                    ($crate::__length_or_infer!(($expr), ($($elem_ty)?),($(const $($len)*)?))),
                    ($([$elem_ty; $($len)*])?),
                )
            )
            ($($($rem)*)?)
        }
    };

    (
        ($($prev:tt)*)
        ( $expr:tt $(: $type:ty)?  $(, $($rem:tt)*)? )
    ) => {
        $crate::__concat_arrays_inner!{
            (
                $($prev)*
                (
                    $expr,
                    ($(<$type as $crate::__::GetTypeParam>::T)?),
                    ($crate::__length_or_infer!(($expr), (), $((type $type))?)),
                    ($($type)?),
                )
            )
            ($($($rem)*)?)
        }
    };

    ( $prev:tt ( $expr:path $(: $($rem:tt)*)? ) ) => {
        $crate::__concat_arrays_inner!{
            $prev
            (($expr) $(: $($rem)*)?)
        }
    };

    ( $prev:tt ( $expr:path $(, $($rem:tt)*)? ) ) => {
        $crate::__concat_arrays_inner!{
            $prev
            (($expr) $(, $($rem)*)?)
        }
    };

}

#[doc(hidden)]
#[macro_export]
macro_rules! __get_array_length {
    ($e:expr; $len:expr) => ({
        $len
    });
    () => (0);
    ($($e:expr),+ $(,)*) => ({
        [$($crate::__as_unit!($e),)*].len()
    });
}

#[doc(hidden)]
#[macro_export]
macro_rules! __length_or_infer {
    (($expr:expr), ($($elem_ty:ty)?), ($(const _)?)) => ({
        let len = $crate::__::Usize;
        if false {
            len.infer_mda $(::<$elem_ty>)? (&$crate::__::ManuallyDrop::new($expr));
        }
        len.get()
    });
    ($expr:tt, $elem_ty:tt, (const $length:expr)) => ( $length );
    ($expr:tt, $elem_ty:tt, (type $type:ty)) => ( <$type as $crate::__::ArrayLength>::LENGTH );
}

#[doc(hidden)]
#[macro_export]
macro_rules! __as_unit {
    ($($tt:tt)*) => {
        ()
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! __type_ascription {
    (($e:expr) ()) => {
        $e
    };
    (($e:expr) ([$elem_ty:ty; _])) => {
        $crate::__::ArrayAndGhost {
            inner: $e,
            elem_ty: $crate::__::PhantomData::<$elem_ty>,
        }
        .inner
    };
    (($e:expr) ($ty:ty)) => {
        $crate::__::Identity::<$ty> { inner: $e }.inner
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! __declare_length_type_and_pass {
    (($length_type:ty), $length:expr) => {{
        impl $length_type {
            pub const LEN: $crate::__::usize = $length;
        }

        <$length_type>::LEN
    }};
    ((), $length:expr) => {
        $length
    };
}


src/lib.rs
//! Array concatenation
//!
//! This crate allows concatenating multiple arrays of varying lengths into one array.
//!
//! # Example
//!
//! ### `concat_arrays`
//!
//! For more examples of using [`concat_arrays`],
//! you can [look here][concat_arrays_examples].
//!
//! ```rust
//! use arrcat::concat_arrays;
//!
//! {
//!     const PRIMES: [u16; 4] = [7, 11, 13, 17];
//!     assert_eq!(
//!         concat_arrays!([3, 4, 4u16.pow(3)], PRIMES),
//!         [3, 4, 64, 7, 11, 13, 17],
//!     );
//! }
//!
//! {
//!
//!     let increasing = [8, 9, 10];
//!
//!     let concated = concat_arrays!(
//!         // the macro can't infer the length of runtime array non-literals.
//!         increasing: [_; 3],
//!         // most non-literal arguments need to be wrapped in `()` or `{}`.
//!         ([2u16, 3, 4].map(|x| x * 9)): [_; 3],
//!     );
//!
//!     assert_eq!(concated, [8, 9, 10, 18, 27, 36]);
//! }
//!
//! ```
//!
//! # No-std support
//!
//! `arrcat` is `#![no_std]`, it can be used anywhere Rust can be used.
//!
//! # Minimum Supported Rust Version
//!
//! `arrcat` requires Rust 1.57.0, requiring crate features to use newer language features.
//!
//!
//! [concat_arrays_examples]: crate::concat_arrays#examples

#![no_std]

#[cfg(test)]
mod tests;

mod internals;

#[macro_use]
mod macros;

#[doc(hidden)]
pub mod __ {
    pub use core::{
        compile_error, concat, marker::PhantomData, mem::ManuallyDrop, primitive::usize, stringify,
    };

    pub use crate::internals::*;
}

#[cfg(feature = "derive")]
#[doc = include_str!("../README.md")]
pub struct ReadmeTest;
